src/lib/middleware/with-access-token.ts

```diff
@@ -14,53 +15,99 @@ export const extractShortToken = (token: string) => token.split("_")?.[1]
+        type: RequiresWorkspaceId extends true
+          ? Extract<AuthenticatedRequest["auth"], { type: "access_token" }>
+          : Extract<
+              AuthenticatedRequest["auth"],
+              { type: "access_token_without_workspace" }
+            >
+      }
+    },
+    {
+      db: Database
     }
-  },
-  {
-    db: Database
-  }
-> = (next) => async (req, res) => {
-  const token = req.headers.authorization?.split("Bearer ")?.[1]
-  if (token == null) return res.status(401).end("Unauthorized")
+  > =>
+  (next) =>
+  async (req, res) => {
+    const token = req.headers.authorization?.split("Bearer ")?.[1]
+    if (token == null) return res.status(401).end("Unauthorized")
 
-  const is_at = token.includes("seam_at")
-  const long_token = token.split("_")?.[2]
-  const short_token = token.split("_")?.[1]
+    const is_at = token.includes("seam_at")
+    const long_token = token.split("_")?.[2]
+    const short_token = token.split("_")?.[1]
 
-  if (short_token == null || long_token == null)
-    return res.status(400).end("malformed token")
+    if (!is_at) {
+      throw new UnauthorizedException({
+        type: "unauthorized",
+        message: "Access tokens must start with seam_at",
+      })
+    }
 
-  if (is_at) {
-    const long_token_hash = extractLongTokenHash(token)
+    const workspace_id = req.headers["seam-workspace"]
 
-    const access_token = req.db.access_tokens.find(
-      (token) => token.long_token_hash === long_token_hash,
-    )
-    if (access_token == null)
+    if (
+      (typeof workspace_id === "undefined" || workspace_id.length === 0) &&
+      require_workspace_id
+    ) {
       throw new UnauthorizedException({
-        type: "access_token_not_found",
-        message: "Access token not found",
+        type: "missing_workspace_id",
+        message: "Workspace ID is required",
       })
-
-    req.auth = {
-      auth_mode: "access_token",
     }
-    // Cannot run middleware after auth middleware.
-    // UPSTREAM: https://github.com/seamapi/nextlove/issues/118
-    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
-    return withSimulatedOutage(next as unknown as any)(
+
+    if (short_token == null || long_token == null)
+      return res.status(400).end("malformed token")
+
+    if (is_at) {
+      const long_token_hash = extractLongTokenHash(token)
+
+      const access_token = req.db.access_tokens.find(
+        (token) => token.long_token_hash === long_token_hash,
+      )
+      if (access_token == null)
+        throw new UnauthorizedException({
+          type: "access_token_not_found",
+          message: "Access token not found",
+        })
+
+      if (require_workspace_id) {
+        ;(req.auth as unknown as Extract<
+          AuthenticatedRequest["auth"],
+          { type: "access_token" }
+        >) = {
+          type: "access_token",
+          workspace_id: workspace_id as string,
+        }
+      } else {
+        ;(req.auth as unknown as Extract<
```

Comment:
Anyway to do this without the `;`? Whenever I see this my brain gets tripped up trying to parse out the nesting.